### ISSUE TITLE
Language tag interpretation

### ISSUE THREAD

**Issue by ptr727** [MediaArea/MediaAreaXml#33]
Hi, another change I observed in the new XML format is the language is no longer in long form, it is either in xy or xy-ab format, e.g. en, or en-us.

This makes conversion a bit tricky, as it is a culture, i.e. language and spelling and units, vs. say just an ISO 639 3 character format that represents the spoken language, e.g. eng.
MKVTools and FFProbe use the ISO-639-2-T 3 letter format.
And, "und" for undetermined, "zxx" for no linguistic content, special handling for "chi" and "zho", because Matroska always uses "chi".

The old XML format used the english long form name, e.g. English or Chinese, and converting that was very reliable.

Can you clarify the language tag spec, and how it will behave for no language tag set (empty?), vs. zxx for no linguistic content (e.g. a video stream), and how variant culture languages like Chinese will be set?

**Comment by ptr727** [comment]
Please understand that with your knowledge my one issue may becomes multiple issues, in my mind it was one question, the language changed from long form to some short form, what is the spec used.
I will try to keep it to one issue per ticket.

You state that you standardize output as ISO 639-3, that is not what I experienced fro my samples.
ISO 639-3 is an expansion of ISO 639-2 B / T types, and all output is expected to be 3 letters, and honour the special designations, like und and zxx:
https://en.wikipedia.org/wiki/ISO_639-3

The output from my sample files produce Language Culture Names, not ISO 639-2 or 639-3 3 letter language code, e.g.
ffprobe and mkvmerge reports "eng", mediainfo reports "en", or "en-us"

If you are standardizing on ISO 639-3, then that is how I will treat the data, it makes my life easier since 99% of the mappings will match ISO-639-2/T which is used by ffmpeg and MKVMerge.
I expect to always get 3 letters, "+ dash + country code, sometimes" in the ISO 639-3 standard.

MKVMerge treats a missing language tag as und, and a language tag can be removed from a MKV by by setting it to und, in which case the tag will be removed. The standard does says to interpret a missing tag as english, but this is intended for players, not for metadata tools.
E.g. https://github.com/mbunkus/mkvtoolnix/issues/1929

You are free to define that you will output no language or und, but it needs to be predictable.

In summary, 
- per your definition, output is expected to be 3 letter ISO 639-3, actual output is language culture info.
- there is no  "+ dash + country code, sometimes" in the ISO 639-3 standard, output is expected to be 3 letters only, (or no output if you define it as such)

**Comment by JeromeMartinez** [comment]
> ou state that you standardize output as ISO 639-3, that is not what I experienced from my samples.
ISO 639-3 is an expansion of ISO 639-2 B / T types, and all output is expected to be 3 letters,

I talked too quickly.
[2 letter codes if exists, else 3-letter code, see full description](
https://github.com/MediaArea/MediaInfoLib/blob/69f86b84951c2bd56392d6d00d5ca0c7567642a7/Source/Resource/Text/Stream/Audio.csv#L256)

> and honour the special designations, like und and zxx

I'll be happy to receive funding or patches for honoring the special designations, in the meanwhile MediaInfo is provided as is (with what I have coded because I needed it or what a sponsor needed).

> ffprobe and mkvmerge reports "eng", mediainfo reports "en", or "en-us"

If you want ffprobe and mkvmerge reports, it is safer for you to use ffprobe and mkvmerge.
Note that the old XML was not better about that (actually worse, no mapping possible).
I did not decide same as FFmpeg (mkvmerge is specific because focuses on a specific format, MediaInfo and FFmpeg are more similar because they have to deal with different inputs), that does not mean that one is worse than the other.
If you have specific needs like "3-letter code only" or "I want to be close to FFmpeg output" or "I want ISO-639-2/T", or "I want a good and precise documentation", you need specific support, see [this page](http://mediaarea.net/Services) for how to have it.
IMO current output is not difficult to parse for having 3-letter code from 2-letter code.

> MKVMerge treats a missing language tag as und

I bet not. the spec is clear about that and Mosu said it often: missing language tag means "eng".
And you link a thread which says the opposite of what you said ("a track which doesn't contain a language attribute is English by definition")

> and a language tag can be removed from a MKV by by setting it to und, in which case the tag will be removed. 

Wrong and you link a thread which says the opposite of what you said ("If you want a track with an undefined language, you have to specify it as undefined.").

Please read your link carefully.

> The standard does says to interpret a missing tag as english, but this is intended for players, not for metadata tools.

???
Players or metadata tools are expected to have same behavior.
You asked "how it will behave for no language tag set ", I answered (MediaInfo follows specs and considers that as "eng" for Matroska so "en" for MediaInfo), that's all.
I add that "und" in Matroska will be converted to empty value in MediaInfo.

> You are free to define that you will output no language or und, but it needs to be predictable.

Done, I think.
I am aware that MediaInfo lacks a lot of doc, but for the moment no one is motivated for doing it or sponsoring it, so it is as is.

**Comment by ptr727** [comment]
Please don't get so defensive.

And really do not say that I should pay if I want docs, if you want us to use your code then either help us use it without getting upset, or document it, make it behave like you documented it, and you would not have to interact with me.

I am not criticising, I am asking clarifying questions.
I am experiencing output that is inconsistent with past versions.
I cannot find any documentation that explains how to interpret the output.
What you say is not what I see, you say 2 or 3, I see 2 or 2-2.

**Comment by JeromeMartinez** [comment]
> Please don't get so defensive.

Sorry, a bit stressed for the moment.
But keep in mind that this is normal that different tools have different behavior, and read carefully (e.g. "a language tag can be removed from a MKV by by setting it to und" is really wrong) else it is difficult to read.

> I am experiencing output that is inconsistent with past versions.

But your questions are about a link with MKVmerge or FFmpeg, and previous versions were worse about mapping between MI and these tools.

> What you say is not what I see, you say 2 or 3, I see 2 or 2-2.

I said 2 or 3 for the first part. "2 or 2-2" fits here.
from the link in my comment: "2-letter ISO 639-1 if exists, else 3-letter ISO 639-2, and with optional ISO 3166-1 country separated by a dash if available, e.g. en, en-us, zh-cn"

For having 3 letters in the first part, you need a language not available in 2-letter form.

**Comment by ptr727** [comment]
Thanks for the clarification.

ffmpeg and mkvmerge are really my problems, not yours, I am not asking you to be like them at all, I just find them to be reasonably well documented, and consistent in their use of ISO 639-2 (except for Chinese)
As for MKVMerge and "und", I am 100% certain that by setting the track language to "und" the language tag gets removed, and MediaInfo old XML reports no language set, I use this behavior today.
Plex, a very popular media player, does not honor the MKV rule of no tag then English, many user complaints, so sometimes a spec is of no use when reality dictates practical workarounds.

In my mind when you said 2, I interpreted that as 2, not 2-2, 2-2 is a culture specification.

But anyway, I will treat the output as 2, or 2-2, or 3, and try to convert it to ISO 639-2/T.

**Comment by JeromeMartinez** [comment]
> As for MKVMerge and "und", I am 100% certain that by setting the track language to "und" the language tag gets removed, and MediaInfo old XML reports no language set, I use this behavior today.

Just checked, MediaInfo reports "en" if tag gets removed.
I have doubts that language tag gets removed if you set track language to "und", Mosu explicitly says it "The reason for this behaviour is that the default value for the language attribute is English and this means that a track which doesn't contain a language attribute is English by definition. If you want a track with an undefined language, you have to specify it as undefined"

If you think that MediaInfo reports empty language if no tag in file, please provide a sample file.

Note that I consider the following sentence as true: As for MKVMerge and "und", I am 100% certain that by setting the track language to "und" the language tag is set to "und", and MediaInfo old XML reports no language set, I use this behavior today.
and if your file contains a "und" language tag, I'll yell ;-) (you say "language tag gets removed").

> Plex, a very popular media player, does not honor the MKV rule of no tag then English, many user complaints, so sometimes a spec is of no use when reality dictates practical workarounds.

Plex needs to be fixed because lot of other players honors the spec.

> But anyway, I will treat the output as 2, or 2-2, or 3,

You'll miss the 3-2 case.

**Comment by JeromeMartinez** [comment]
>  MediaInfo old XML reports no language set, I use this behavior today.

As far as I know, there is no change between Old XML and new XML about that.

**Comment by ptr727** [comment]
If I call mkvpropedit and I set a track language to und, that is how I was told to remove the language, then old MI reports no language, MKVMerge --identify and ffprobe will report und, my code tests for no language defined by looking for no tag set in old MediaInfo old XML, or und in MKVMerge --identify.

For the 3-2 case, what is the specification of this format? 
As far as I know there is no culture or ISO spec that follows a 3-2 format, is there, I need to write logic to convert to ISO 639-2.

**Comment by JeromeMartinez** [comment]
> that is how I was told to remove the language

I get it now. by "how to remove the language tag", you mean in MediaInfo, not in the file.
Be careful, "As for MKVMerge and "und", I am 100% certain that by setting the track language to "und" the language tag gets removed" means something different (that the tag in the file is removed because you peak of MKVmerge so about the file produced by MKVmerge, not the result in MediaInfo) and leads to misunderstandings.
And again, no change in MediaInfo between old XML and new XML, so I don't see the reason it is there.

> For the 3-2 case, what is the specification of this format?

?
Same as before: "2-letter ISO 639-1 if exists, else 3-letter ISO 639-2, and with optional ISO 3166-1 country separated by a dash if available, e.g. en, en-us, zh-cn"
"3-2" fits in this explanation.

> As far as I know there is no culture or ISO spec that follows a 3-2 format,

We talk about MediaInfo, don't we? so MediaInfo definition (just above)

> I need to write logic to convert to ISO 639-2.

ISO 639-2 is 3-letter code. As said at the beginning except my mistake in first answer, language is first part (before the dash) so remove dash and after dash if dash is present, and if 2-letter then convert to 3-letter, else it is already ISO 639-2.

**Comment by ptr727** [comment]
I just double checked, if I set a track to und in mkvpropedit, then the language xml tag is missing the the mediainfo output with the old and the new XML format.

For format, sorry, I think I am a bit confused by meaning, let me try some examples.
When you say 3-2, I understand 3 letters dash 2 letters, e.g. eng-us.
That does not look like a standard to me?

Do you mean 3 or 2 or 2-2?
2 = en
2-2 = en-us
3 = eng

If so, then I I know how to parse that.

**Comment by JeromeMartinez** [comment]
> When you say 3-2, I understand 3 letters dash 2 letters, e.g. eng-us.

"eng-us" does **not** fit in the description "2-letter ISO 639-1 if exists, else 3-letter ISO 639-2, and with optional ISO 3166-1 country separated by a dash if available" because 2-letter ISO 639-1 of "eng" exists ("en").

> Do you mean 3 or 2 or 2-2?

No: I mean "2-letter ISO 639-1 if exists, else 3-letter ISO 639-2, and with optional ISO 3166-1 country separated by a dash if available" so 2, 2-2, 3, 3-2 are possible (in theory).
You example is bad because you use "eng" for the 3-letter code and 2-letter ISO 639-1 of "eng" exists ("en").

> That does not look like a standard to me?

I did not specify any standard, I said "2-letter ISO 639-1 if exists, else 3-letter ISO 639-2, and with optional ISO 3166-1 country separated by a dash if available"

so e.g. "ace-in" (because there is no ISO 639-1 code for "ace" ISO 639-2).

***

Theory only but you said you want me to be precise, so I am precise.
I order to be even more precise, I add that if the Matroska (or other) has e.g. 100 letters in the language tag (not "legal" in specs but technically possible), I'll show the 100 letters. Be prepared to have anything.

**Comment by ptr727** [comment]
Ok, I used a bad example :)

I will parse for (where abcde are just example letters, not real countries or languages):
blank or missing tag = ISO 639-2 und
ab = ISO 639-1 convert to ISO 639-2
ab-cd = language culture convert to ISO 639-2
abc = ISO 639-2

I do not know what to do with:
abc-de = assert, get example file, figure it out
abcdefghijklmn = assert, get example file, figure it out

Thank you

**Comment by JeromeMartinez** [comment]
> I do not know what to do with:

Not real use cases, so "passthrough" (no change) and user checks.

**Comment by HughP** [comment]
Sorry to be late to the game here (by about a year) it seems that @ptr727 should be looking at BCP47 for the interpretation of language codes. https://tools.ietf.org/html/bcp47  BCP47 does use ISO 639-3 codes but that is only if there is not a valid IANA language tag. A good article can be found here: https://www.w3.org/International/articles/language-tags/ and the IANA registry can be found here: http://www.iana.org/assignments/language-subtag-registry/language-subtag-registry

And a great look up tool for valid tags can be found here: https://r12a.github.io/app-subtags/

**Comment by ptr727** [comment]
I switched to BCP47 :)
